
# Gene Drive Mutation Accumulation Project
**Michael Overton**
Kryazhimskiy Lab UCSD
   
   
## Project Abstract
CRISPR-Cas9 gene drives (CCGDs) are poised to expand our ability to control wild populations, such as the eradication of disease vectors. However, Cas9-gRNA complexes can produce off-target dsDNA breaks, and this raises the possibility that a CCGD element could increase rates of mutations, genome instability, and loss of heterozygosity (LOH), with unpredictable long-term evolutionary consequences. To assess the mutational burden a CCGD may impose, we performed a large-scale mutation accumulation experiment in a hybrid strain of the yeast Saccharomyces cerevisiae transformed with one of three constructs: an empty control (WT), a Cas9-only element (Cas9), or the full Cas9 + gRNA CCGD element (Drive). Neither de novo mutation nor LOH event rates were statistically distinguishable between the strains. Given the statistical power of our study, we estimate with 70% confidence that the presence of a CCGD element increases the genome-wide rates of LOH events and de novo mutations by less than 25% and 70%, respectively. Despite the lack of significant differences for the genome-wide rates, we found a significant difference in the distribution of LOH events across Chromosome V between the WT and Drive strains. Our results suggest that these elements impose at most a weak and likely localized additional mutational burden to a naive genome, and that CCGD-based population control efforts are likely to be evolutionarily safe.

**[Experimental Design](https://github.com/moverton88/Gene_Drive_MutAccum/blob/master/Experimental_Design.md)**

[ED2](https://github.com/moverton88/Gene_Drive_MutAccum/Experimental_Design.md)

**[Code](https://github.com/moverton88/Gene_Drive_MutAccum/)**

***[Sequencing and Genotyping]()***
